mavros_extras/src/plugins/adsb.cpp

```diff
@@ -0,0 +1,127 @@
+using mavlink::common::ADSB_EMITTER_TYPE;
+using mavlink::common::ADSB_FLAGS;
+
+/**
+ * @brief ADS-B Vehicle plugin
+ *
+ * Publish/subscribe Automatic dependent surveillance-broadcast data to/from a vehicle.
+ */
+class ADSBPlugin : public plugin::PluginBase {
+public:
+	ADSBPlugin() : PluginBase(),
+		adsb_nh("~adsb")
+	{ }
+
+	void initialize(UAS &uas_)
+	{
+		PluginBase::initialize(uas_);
+
+		adsb_nh.param<std::string>("frame_id", frame_id, "adsb_vehicle");
+
+		adsb_pub = adsb_nh.advertise<mavros_msgs::ADSBVehicle>("vehicle", 10);
+		adsb_sub = adsb_nh.subscribe("own", 10, &ADSBPlugin::adsb_cb, this);
+	}
+
+	Subscriptions get_subscriptions()
+	{
+		return {
+			       make_handler(&ADSBPlugin::handle_adsb)
+		};
+	}
+
+private:
+	ros::NodeHandle adsb_nh;
+
+	std::string frame_id;
+
+	ros::Publisher adsb_pub;
+	ros::Subscriber adsb_sub;
+
+	ADSB_ALTITUDE_TYPE altitude_type;
+	ADSB_EMITTER_TYPE emitter_type;
+
+	void handle_adsb(const mavlink::mavlink_message_t *msg, mavlink::common::msg::ADSB_VEHICLE &adsb)
+	{
+		auto adsb_msg = boost::make_shared<mavros_msgs::ADSBVehicle>();
+		const auto &adsb_callsign = adsb.callsign;
+
+		ros::Time tslc(adsb.tslc, 0);
+
+		adsb_msg->header.stamp = ros::Time::now();	//TODO: request add time_boot_ms to msg definition
+		adsb_msg->header.frame_id = frame_id;
+		adsb_msg->ICAO_address = adsb.ICAO_address;
+		adsb_msg->lla.latitude = adsb.lat / 1E7;
+		adsb_msg->lla.longitude = adsb.lon / 1E7;
+		adsb_msg->lla.altitude = adsb.altitude;	// in meters, TODO: #693
+		adsb_msg->altitude_type = adsb.altitude_type;
+		adsb_msg->heading = adsb.heading / 1E2 / 180.0f * M_PI - M_PI;
+		adsb_msg->hor_velocity = adsb.hor_velocity / 1E2;
+		adsb_msg->ver_velocity = adsb.ver_velocity / 1E2;	// up is positive
+		std::copy(adsb_callsign.cbegin(), adsb_callsign.cend(), adsb_msg->callsign.begin());
+		adsb_msg->emitter_type = adsb.emitter_type;
+		adsb_msg->tslc.data = tslc;
+		adsb_msg->data_status = adsb.flags;
+		adsb_msg->squawk = adsb.squawk;
+
+		// for info purposes
+		ROS_DEBUG_NAMED("adsb", "receiving data of altitude type: %s", utils::to_string(adsb.altitude_type).c_str());
+		ROS_DEBUG_NAMED("adsb", "receiving data of emitter type: %s", utils::to_string(adsb.emitter_type).c_str());
+		ROS_DEBUG_NAMED("adsb", "data status flags on receiving msg: %s", utils::to_string(adsb.flags).c_str());
+
+		adsb_pub.publish(adsb_msg);
+	}
+
+	void adsb_cb(const mavros_msgs::ADSBVehicle::ConstPtr &req)
+	{
+		mavlink::common::msg::ADSB_VEHICLE adsb {};
+
+		adsb.ICAO_address = req->ICAO_address;
+		adsb.lat = req->lla.latitude * 1E7;
+		adsb.lon = req->lla.longitude * 1E7;
+		adsb.altitude = req->lla.altitude;	// in meters, TODO: #693
+		adsb.altitude_type = req->altitude_type;
+		adsb.heading = req->heading * 1E2 * 180.0f / M_PI + M_PI;
+		adsb.hor_velocity = req->hor_velocity * 1E2;
+		adsb.ver_velocity = req->ver_velocity * 1E2;	// up is positive
+		mavlink::set_string(adsb.callsign, req->callsign);
```

Comment:
Should be `mavlink::set_string_z()`, _z is important.
`set_string()` is for message fields like param_id, where `'\0'` may be omitted and only indicate shorter strings than buffer.
`set_string_z()` ensures that last byte is `'\0'`, which required by adsb message desc.